### DISCUSSION TITLE
why there is no coupledVectorNeighborValue?

### DISCUSSION THREAD

**Discussion by dingqiushi** [idaholab/moose#22473] [Q&A General] [answered]
I can find coupledNeighborValue() function in moose. But I need to couple vector value in my kernel. Is there any way to achieve this? I have not find fuction coupledVectorNeighborValue.

**Comment by GiudGiud** [comment, marked as answer]
Hello

So you are using a vector variable?
I'm not sure why we dont have that one but we have:
```
adCoupledVectorNeighborValue
```
and you can always retrieve the value from an AD number

Guillaume

**Reply by dingqiushi** [reply]
Thank you. 
I am using vector variable. But not AD number. I find ` CZMMaterialBase` retrieve neighbor displacements vector value like:
```
InputParameters
CZMMaterialBase::validParams()
{
  InputParameters params = InterfaceMaterial::validParams();

  params.addClassDescription("Base class for cohesive zone mateirla models");
  params.addRequiredCoupledVar("displacements",
                               "The string of displacements suitable for the problem statement");
  return params;
}

CZMMaterialBase::CZMMaterialBase(const InputParameters & parameters)
  : InterfaceMaterial(parameters),
    _normals(_assembly.normals()),
    _ndisp(coupledComponents("displacements")),
    _disp(_ndisp),
    _disp_neighbor(_ndisp),
    _displacement_jump_global(declareProperty<RealVectorValue>("displacement_jump_global")),
    _displacement_jump(declareProperty<RealVectorValue>("displacement_jump")),
    _traction_global(declareProperty<RealVectorValue>("traction_global")),
    _traction(declareProperty<RealVectorValue>("traction")),
    _traction_derivatives_global(declareProperty<RankTwoTensor>("traction_derivatives_global")),
    _traction_derivatives(declareProperty<RankTwoTensor>("traction_derivatives"))
{
  if (_ndisp > 3 || _ndisp < 1)
    mooseError("the CZM material requires 1, 2 or 3 displacement variables");

  if (getParam<bool>("use_displaced_mesh") == true)
    mooseError("This material cannot be used with use_displaced_mesh = true");

  // initializing the displacement vectors
  for (unsigned int i = 0; i < _ndisp; ++i)
  {
    _disp[i] = &coupledValue("displacements", i);
    _disp_neighbor[i] = &coupledNeighborValue("displacements", i);
  }
}
```
Can I use the same method to couple neighbor vector value in my kernel?

**Reply by GiudGiud** [reply]
Yes, it s ok to retrieve the AD coupled value. When you use it you can do var.value() to retrieve the number

Either that or you add this coupledVectorNeighborValue routine to MOOSE. It wont be too hard since the AD version exists already

**Reply by hugary1995** [reply]
There might be a misunderstanding in terminology. Displacement is a vector, yes -- that's how I would say that in maths... But in MOOSE, a "vector variable" has a very specific meaning. A vector variable in MOOSE is a variable that has exactly three components, for example of family `LAGRANGE_VEC`. There are then apparently two ways of defining the displacement variables in MOOSE:

1. As a vector variable:
```python
[disp]
  order = FIRST
  family = LAGRANGE_VEC
[]
```

2. As three variables:
```python
[disp_x]
  order = FIRST
  family = LAGRANGE
[]
[disp_y]
  order = FIRST
  family = LAGRANGE
[]
[disp_z]
  order = FIRST
  family = LAGRANGE
[]
```

In the tensor_mechanics module, we always do option 2, in which case we need to loop over the components like in the code you pasted above:
```cpp
for (unsigned int i = 0; i < _ndisp; ++i)
    _disp[i] = &coupledValue("displacements", i);
```
or use an equivalent shortcut:
```cpp
_disp = coupledValues("displacements");
```
So you'll want to do something similar if you choose option 2.

You need `coupledVectorValue` and/or `coupledVectorNeighborValue` only if you are doing option 1.
